### PR TITLE
`topk` on CUDA supports `bfloat16`

### LINKED ISSUE #56176: TopK only implements partial support for BFloat16
## 🐛 Bug

While porting topK from THC to ATen (#55392), we noticed that only the sorting codepath for smaller tensors (<= 2048 elements) supports BFloat16. So while topk ostensibly supports BFloat16, passing an input tensor with more than 2048 elements will raise a not implemented error. 

## To Reproduce

1. `a = torch.randn(3072, dtype=torch.bfloat16).cuda()`
2. `a.topk(5)`

(Compare with smaller input size <= 2048).

## Expected behavior

See above.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

 - PyTorch Version (e.g., 1.0): 1.8.0 (pulled from recent upstream)
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source):
 ```
 TORCH_CUDA_ARCH_LIST="7.0" CUDA_HOME="/usr/local/cuda"     CMAKE_PREFIX_PATH="$(dirname $(which conda))/../"     NCCL_INCLUDE_DIR="/usr/include/"     NCCL_LIB_DIR="/usr/lib/"     USE_SYSTEM_NCCL=1 USE_OPENCV=1 python setup.py develop
 ```
 - Python version: 3.6.10
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: Tesla T4
 - Any other relevant information:
 
 CC @zasdfgbnm @ezyang

### PR DESCRIPTION
**Pull request by eqy**
Fixes #56176 via #58196

CC @zasdfgbnm @ngimel @ptrblck

### CODE REVIEW

**Comment by eqy** [comment]
running CI to see what happens on ROCM, probably need to add skips for the new tests

**Comment by ngimel** [comment]
ROCm is failing as expected, you need to adjust skips.

File: test/test_sort_and_select.py [hunk 1]
```diff
@@ -581,21 +581,42 @@ def test_topk_noncontiguous_gpu(self, device):
...
-            topk = a.topk(curr_size // 2)
-            self.assertEqual(sort_topk, topk[0])      # check values
-            self.assertEqual(sort_topk, a[topk[1]])   # check indices
+            self._test_topk_dtype(device, dtype, True, curr_size)
+
+    @onlyCUDA
+    @dtypes(torch.bfloat16)
+    def test_topk_bfloat16(self, device, dtype):
+        if TEST_WITH_ROCM and dtype == torch.bfloat16:
+            return
```
Comment (ngimel) [inline test/test_sort_and_select.py:607]:
```suggestion
    @onlyCUDA
    @dtypes(torch.bfloat16)
    @skipCUDAIfRocm
    def test_topk_bfloat16(self, device, dtype):
```

**Review by ngimel — APPROVED** [review]

File: test/test_sort_and_select.py [hunk 2]
```diff
@@ -581,21 +581,41 @@ def test_topk_noncontiguous_gpu(self, device):
...
+    @dtypes(torch.bfloat16)
+    @skipCUDAIfRocm
+    def test_topk_bfloat16(self, device, dtype):
+
+        small = 10
+        large = 8192
+        for curr_size in (small, large):
+            self._test_topk_dtype(device, dtype, False, curr_size)
 
     @dtypesIfCUDA(*torch.testing.get_all_fp_dtypes())
```
Comment (ngimel) [inline test/test_sort_and_select.py:613]: you probably should `include_bfloat16=True` here, otherwise it's not tested

**Pull request closed without merging**